crates/nu-protocol/src/engine/stack.rs

```diff
@@ -211,12 +211,13 @@ impl Stack {
     ///
     /// The config will be updated with successfully parsed values even if an error occurs.
     pub fn update_config(&mut self, engine_state: &EngineState) -> Result<(), ShellError> {
-        if let Some(mut config) = self.get_env_var(engine_state, "config") {
-            let existing_config = self.get_config(engine_state);
-            let (new_config, error) = config.parse_as_config(&existing_config);
-            self.config = Some(new_config.into());
+        if let Some(value) = self.get_env_var(engine_state, "config") {
+            let old = self.get_config(engine_state);
+            let mut config = (*old).clone();
+            let error = config.update_from_value(&old, &value);
             // The config value is modified by the update, so we should add it again
-            self.add_env_var("config".into(), config);
+            self.add_env_var("config".into(), config.clone().into_value(value.span()));
+            self.config = Some(config.into());
```

Comment:
OK, that is probably a performance regression. 

Currently we clone the `Value` representation of the config once in `Stack::get_env_var` (somewhat unfortunate), then the existing `Value::parse_as_config` simply mutates the existing nested structure as needed for error correction. The final Value is then moved into the `$env.config`.

In the new version the `Config` struct is now also cloned and then all the nested records have to be allocated again by `.into_value()`.

Reply:
If we changed `get_env_var` to return a reference to a value, then I think the performance would be similar (one value clone and one config allocation vs two config clones).